share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -240,9 +249,87 @@ public void startup(
 
         log.info("Starting up.");
         numPartitions = shareGroupTopicPartitionCount.getAsInt();
+        setupRecordPruning();
         log.info("Startup complete.");
     }
 
+    // visibility for tests
+    void setupRecordPruning() {
+        log.info("Scheduling share state topic prune job.");
+        timer.add(new TimerTask(config.shareCoordinatorTopicPruneIntervalMs()) {
+            @Override
+            public void run() {
+                List<CompletableFuture<Void>> futures = new ArrayList<>();
+                runtime.activeTopicPartitions().forEach(tp -> futures.add(performRecordPruning(tp)));
+
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[]{}))
+                    .whenComplete((res, exp) -> {
+                        if (exp != null) {
+                            log.error("Received error in share state topic prune, stopping job.", exp);
+                            return;
+                        }
```

Comment:
Are you sure about this? Don't you want to retry after the next interval?

Reply:
@dajac 
But you mentioned:
```
I also wonder whether we should re-reschedule in case of failure too. thenRun would only reschedule if the future succeeds. It would make it more robust to unexpected failures.
```

I have modified the `performRecordPruning` to complete the future exceptionally in cases
- state machine lookup throws error OTHER THAN NOT_COORDINATOR or COORDINATOR_LOAD_IN_PROGRESS
- `writer.deleteRecords` throws exception
- other exception cases like  NOT_COORDINATOR or COORDINATOR_LOAD_IN_PROGRESS - future completes normally. Abnormal offset value also it completes normally.